### PR TITLE
Add support for WebUSB

### PR DESCRIPTION
**Pull request by drashna**
## Description

This adds the WebUSB feature to QMK Firmware (for LUFA and ChibiOS, at least). 

This requires qmk/lufa#8 to be merged (or to use the web_usb)

## Types of Changes

- [x] Core
- [x] New feature
- [ ] Documentation

## Checklist
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by yulei** [comment]
great, this made online qmk configurator possible :)

**Comment by drashna** [comment]
@yulei ZSA (ergodox ez, planck ez) are already using WebUSB for the "live training" feature.  In fact, this is basically just backporting that code.

**Comment by yulei** [comment]
> @yulei ZSA (ergodox ez, planck ez) are already using WebUSB for the "live training" feature. In fact, this is basically just backporting that code.

Could this webusb implementation work under mac? I tested an implementation from the LUFA pr before, but they failed work under MAC OS.

**Comment by drashna** [comment]
Works on MacOS, Windows, and Linux.  including the "open link" popup.  But it does require Chrome or a chrome based browser.

File: keyboards/ergodox_ez/config.h [hunk 1]
```diff
@@ -25,9 +25,10 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define VENDOR_ID       0xFEED
 #define PRODUCT_ID      0x1307
 #define DEVICE_VER      0x0001
-#define MANUFACTURER    ErgoDox EZ
-#define PRODUCT         ErgoDox EZ
+#define MANUFACTURER    ZSA
+#define PRODUCT         Ergodox EZ
```
Comment (noroadsleft) [inline keyboards/ergodox_ez/config.h]: https://ergodox-ez.com/ seems to use a capital D throughout.

**Comment by wiserfirst** [comment]
@drashna @noroadsleft Just wondering if we plan to merge this in? Thanks!

**Comment by noroadsleft** [comment]
@wiserfirst My understanding is that ZSA (makers of the ErgoDox EZ and Planck EZ) is still working on this, hence this PR being in Draft status.

I would expect this to be merged when the underlying pieces are in place. So basically: "When it's done."

**Comment by drashna** [comment]
> Just wondering if we plan to merge this in? Thanks!

It's complicated. 

This PR requires a change to the LUFA module first. Which hasn't happened. Without that change, this PR will fail (right now, it's pointing at the ZSA LUFA module, which is the ONLY reason it works). 

Additionally, we (ZSA) haven't had a chance to really test this out yet.  The WebUSB part works, but the rest of it isn't tested, and we'd rather have that finished before this gets merged. 

That said, once things are finalized on our side, yes, we plan on finishing up this PR and getting this merged.

Until then, we'd rather it remain as a draft.

**Comment by wiserfirst** [comment]
@drashna awesome. Thanks for the detailed explanation!

> right now, it's pointing at the ZSA LUFA module, which is the ONLY reason it works

Can I assume that the zsa fork (https://github.com/zsa/qmk_firmware) would support WebUSB? Thanks!

**Comment by drashna** [comment]
Sorry, I meant to respond sooner!

> Can I assume that the zsa fork ([zsa/qmk_firmware](https://github.com/zsa/qmk_firmware)) would support WebUSB? Thanks!

Yes, it does.  And ZSA is using it for the live training mode, actively. 

And I just added a PR to QMK so that it doesn't have the compile errors when using a keymap for Oryx.

File: quantum/oryx.c [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            uint16_t size   = param[2];  // size <= 28
+            uint8_t event[size+3];
+            uint8_t i;
+            dynamic_keymap_get_buffer(offset, size, &param[3]);
+            event[0] = WEBUSB_STATUS_OK;
+            event[1] = ORYX_EVT_LIVE_UPDATE_GET_BUFFER;
+            for (i = 0; i < size; i++) {
+                event[i+2] = param[i];
+            }
+            event[i+2] = WEBUSB_STOP_BIT;
```
Comment (LSChyi) [inline quantum/oryx.c:81]: Why not use:

```suggestion
            dynamic_keymap_get_buffer(offset, size, &event[2]);
            event[0] = WEBUSB_STATUS_OK;
            event[1] = ORYX_EVT_LIVE_UPDATE_GET_BUFFER;
            event[size+2] = WEBUSB_STOP_BIT;
```

? The buffer is already allocated. Also, I think there is a bug in the loop of copying things back.

**Comment by drashna** [comment]
Honestly, I think that the direction that QMK is going with is webHID, which uses the raw HID endpoint.   As such, I'm not sure how much use this has, and maintaining this isn't worth it.

**Comment by RuRo** [comment]
@drashna can you clarify, what this means for the live training support? You have [previously implied](https://www.reddit.com/r/olkb/comments/g5yhvg/contribution_of_zsa_to_qmk/) that you were planning to backport webUSB and live training stuff to qmk. Now that WebUSB is apparently being abandoned, what are your plans regarding this in the future?

**Comment by drashna** [comment]
Likely, a pivot to webHID.  WebHID uses the raw HID endpoint via the web browser.  Chrome added support for webHID 2 or so years ago, and it is supported by any chromium based browser. 

This would be a better way to go, as it would work with raw HID enabled apps like VIA, but also for online based apps (such as [Remapper](https://remap-keys.app/).  

And QMK XAP is planning on using raw HID, and we'd like to support that.

**Comment by RuRo** [comment]
Ah, okay, thank you for the quick answer. Last question. Just to clarify, will the zsa fork and the configure.zsa.io/train live training app also eventually move to webHID?

**Comment by drashna** [comment]
> Ah, okay, thank you for the quick answer. Last question. Just to clarify, will the zsa fork and the configure.zsa.io/train live training app also eventually move to webHID?

I can't say for sure, but I believe that's the goal long term. Especially when QMK XAP is finalized.

**Pull request closed without merging**